### ISSUE TITLE
Printing to stdout does not work when calling eta code from leinigen

### ISSUE THREAD

**Issue by JustusAdam** [typelead/eta#256]
My minimal example is the following:

The eta code:

```Haskell
data {-# CLASS "MyClass" #-} MyClass = MyClass (Object# MyClass)
printStuff :: Java MyClass ()
printStuff =
    io $ do 
        putStrLn "To stdout"
        hPutStrLn stdout "To stdout"
        hPutStrLn stderr "To stderr"
```

The clojure code:
```clojure
(ns debug-ns (:import MyClass))

(.printStuff (MyClass.))
```

I use the leinigen command `lein run -m debug-ns`.

And the result:
```
To stderr
```
(plus a leinigen error that it doesn't find anything to run, but that doesn't actually matter.

FYI running the code 

```Haskell
printStuff2 = do
    putStrLn "To stdout"
    hPutStrLn stdout "To stdout"
    hPutStrLn stderr "To stderr"
```

with `epm run` does work as intended, however the stderr stuff appears before the stdout stuff:

```
Running <executable>...
To stderr
To stdout
To stdout
```

**Comment by rahulmutt** [comment]
Can you try running `MyClass.printStuff()` directly with `java` and share the result?

**Comment by JustusAdam** [comment]
Same problem. Also only prints the stderr.

**Comment by pparkkin** [comment]
Here are the two files I'm using for testing.

`Test.hs`
```
{-# LANGUAGE MagicHash #-}

module Test where

import System.IO ( hPutStrLn
                 , hFlush
                 , stdout
                 , stderr
                 )
import Java

data {-# CLASS "java_interop.TestStdout" #-} TestStdout
  = TestStdout (Object# TestStdout)
  deriving Class

testStdout :: Java TestStdout ()
testStdout = io $ do
  hPutStrLn stderr "(stderr) 1. from Eta (hPutStrLn)"
  putStrLn "(stdout) 2. from Eta (putStrLn)"
  -- hFlush stdout
  hPutStrLn stdout "(stdout) 3. from Eta (hPutStrLn)"
  -- hFlush stdout
  hPutStrLn stderr "(stderr) 4. from Eta (hPutStrLn)"
  -- hFlush stdout

foreign export java testStdout :: Java TestStdout ()
```

`Main.java`
```
import java_interop.TestStdout;
import eta.runtime.Rts;

public class Main {

  public static void main(String[] argv) {
    TestStdout test = new TestStdout();
    System.out.println("(stdout) Before call into Eta code.");
    test.testStdout();
    System.out.println("(stdout) After call into Eta code.");
    System.out.println("(stdout) Calling Rts.hsExit().");
    Rts.hsExit();
  }
}
```

First thing is, you need to call `Rts.hsExit()` to flush `stdout`. I'm not yet sure why `stderr` behaves differently. By the way, even if you call `Rts.hsExit()` right after calling into the Eta function, the output isn't interleaved like it is in the function. All the `stderr` comes out during the call, and all the `stdout` comes out only when the call to `Rts.hsExit()` is made.

Here's the output I get from running these.
```
(stdout) Before call into Eta code.
(stderr) 1. from Eta (hPutStrLn)
(stderr) 4. from Eta (hPutStrLn)
(stdout) After call into Eta code.
(stdout) Calling Rts.hsExit().
(stdout) 2. from Eta (putStrLn)
(stdout) 3. from Eta (hPutStrLn)
```

What you can see here is that anything that goes into `stdout` gets buffered until `Rts.hsExit()` is called. But the weird thing is that you can write to `stdout`, and it gets output before everything that's flushed in `Rts.hsExit()`.

It's like everything going to `stdout` is getting buffered somewhere other than `stdout`, and anything going directly to `stdout` goes through, and then finally when `Rts.hsExit()` is called, everything in the other buffer gets flushed.

**Comment by rahulmutt** [comment]
I think this has to do with how IO is implemented. The default is to use `BlockBuffering` so there's a block buffer of size 8KB (see `GHC.IO.Handle.Internals` in `base`) that is allocated for the purposes of I/O on the Eta side that has no relation with the buffer in `System.out` that is flushed on call to `println`. That block buffer will only be flushed when:
1. The 8KB buffer is filled (will not happen for short programs like above)
2. `hFlush` is invoked on the handle
3. The runtime exits and `flushStdHandles` is called that flushes `stdout` and `stderr`.

While we're on the topic, I'm wondering whether making `LineBuffering` the default is better. `BlockBuffering` is a real pain and often unintuitive for users. In your program above, run `hSetBuffering LineBuffering` at the top to see what the behaviour is. Thoughts?

**Comment by rahulmutt** [comment]
But this doesn't explain why the same thing isn't happening for `stderr` - `stderr` should also be `BlockBuffering` by default and hence shouldn't have printed when called from Java without `hsExit()`.

**Comment by rahulmutt** [comment]
The answer is in the definition of the `stderr` `Handle`:

```haskell
-- | A handle managing output to the Haskell program's standard error channel.
stderr :: Handle
{-# NOINLINE stderr #-}
stderr = unsafePerformIO $ do
    -- ToDo: acquire lock
   setBinaryMode FD.stderr
   enc <- getLocaleEncoding
   mkHandle FD.stderr "<stderr>" WriteHandle False{-stderr is unbuffered-}
                (Just enc)
                nativeNewlineMode{-translate newlines-}
                (Just stdHandleFinalizer) Nothing
```

Note that `{-stderr is unbuffered-}`. Following the definition in `mkHandle`, it turns out the the `Handle` has `NoBuffering` set on it, so every write will flush the buffer.

**Comment by pparkkin** [comment]
Adding `hSetBuffering stdout LineBuffering` fixes the issue, and the output is what you would expect.
```
(stdout) Before call into Eta code.
(stderr) 1. from Eta (hPutStrLn)
(stdout) 2. from Eta (putStrLn)
(stdout) 3. from Eta (hPutStrLn)
(stderr) 4. from Eta (hPutStrLn)
(stdout) After call into Eta code.
(stdout) Calling Rts.hsExit().
```

Personally, I think line buffering for `stdout` is probably a less surprising default than an arbitrary sized block buffer.

**Comment by rahulmutt** [comment]
Yeah even I feel the same. Would you be willing to submit a patch to set `LineBuffering` to default for `stdout`?

**Comment by rahulmutt** [comment]
Or even `NoBuffering`, depending on what seems to be better.

**Comment by pparkkin** [comment]
Sure thing.

**Comment by pparkkin** [comment]
https://github.com/typelead/eta/pull/419

**Comment by jneira** [comment]
Closed by #419 
(But it seems to be somewhat related with #710)
